src/lib.rs

```diff
@@ -84,19 +110,22 @@ use std::time::Duration;
 pub mod delay;
 #[cfg(feature = "async")]
 pub mod async;
+pub mod opresult;
+
+pub use opresult::OperationResult;
 
 /// Retry the given operation synchronously until it succeeds, or until the given `Duration`
 /// iterator ends.
 pub fn retry<I, O, R, E>(iterable: I, mut operation: O) -> Result<R, Error<E>>
-where I: IntoIterator<Item=Duration>, O: FnMut() -> Result<R, E> {
+where I: IntoIterator<Item=Duration>, O: FnMut() -> OperationResult<R, E> {
```

Comment:
How about having `O` return a generic type that must implement `Into<OperationResult>`? Then the user won't have to call `.into()` explicitly at the call site.

Reply:
Oh! is that how to do it, I had a google around but didn't find a clear answer.

So just replace OperationResult<R, E> with `Into<OperationResult<R,E>>`? 

I've pushed up the rebase, will test out that now.

Reply:
ah hmm, Into is a trait, so thats unsized, 'impl trait' isn't allowed on closure types (E0562);. I'm very happy to do it, but may need some guidance - my rust is rapidly improving but corners like this still take some time for me to navigate.

Reply:
It'll work with the return type as a bounded generic, i.e.:

``` rust
pub fn retry<I, O, R, E, OR>(iterable: I, mut operation: O) -> Result<R, Error<E>>
where
    I: IntoIterator<Item = Duration>,
    O: FnMut() -> OR,
    OR: Into<OperationResult<R, E>>,
```

Then you can change `match operation()` to `match operation().into()`, which will let you get rid of all the other `into` calls. (Remember to reword the docs mentioning that, too.)

Reply:
yeah, the rest of the golf was clear; adding the bound to the fn list of template parameters is what I didn't know to do - thank you! Pushed up that now.